### PR TITLE
Add cluster time skew warning

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: lxd/cluster/gateway.go [hunk 1]
```diff
@@ -196,19 +197,37 @@ func (g *Gateway) HandlerFuncs(nodeRefreshTask func(*APIHeartbeat), trustedCerts
...
+					logger.Warn("Time skew detected between leader and local", log.Ctx{"leaderTime": heartbeatData.Time, "localTime": now})
+
+					err := g.Cluster.UpsertWarning("", "", -1, -1, db.WarningClusterTimeSkew, fmt.Sprintf("leaderTime: %s, localTime: %s", heartbeatData.Time, now))
+					if err != nil {
+						logger.Warn("Failed to create cluster time skew warning")
+					}
 				}
 				g.timeSkew = true
-			} else if heartbeatData.Time.Add(-5 * time.Second).After(time.Now().UTC()) {
+			} else if heartbeatData.Time.Add(-5 * time.Second).After(now) {
```
Comment (tomponline) [inline lxd/cluster/gateway.go:212]: @monstermunchkin could you merge this stanza condition with the one above as an OR statement to reduce the duplication here?

File: lxd/cluster/gateway.go [hunk 2]
```diff
@@ -196,19 +197,27 @@ func (g *Gateway) HandlerFuncs(nodeRefreshTask func(*APIHeartbeat), trustedCerts
...
-				g.timeSkew = true
-			} else if heartbeatData.Time.Add(-5 * time.Second).After(time.Now().UTC()) {
+			now := time.Now().UTC()
+
+			if heartbeatData.Time.Add(5*time.Second).Before(now) || heartbeatData.Time.Add(-5*time.Second).After(now) {
 				if !g.timeSkew {
-					logger.Warn("Time skew detected between leader and local", log.Ctx{"leaderTime": heartbeatData.Time, "localTime": time.Now().UTC()})
+					logger.Warn("Time skew detected between leader and local", log.Ctx{"leaderTime": heartbeatData.Time, "localTime": now})
+
+					err := g.Cluster.UpsertWarning("", "", -1, -1, db.WarningClusterTimeSkew, fmt.Sprintf("leaderTime: %s, localTime: %s", heartbeatData.Time, now))
```
Comment (tomponline) [inline lxd/cluster/gateway.go:206]: @monstermunchkin @stgraber more of a passing though really, but maybe the `message` arg for `UpsertWarning` would be more useful as a structured argument (e.g. a `map[string]string` or a `log.Ctx` type) as it seems to be being used mostly to put contextual info into the warning. We could then encoded that as JSON when its stored in the DB and decode when returned by the API? That would probably mean an API change, but seeing as this feature is relatively new and unused, if that is of any value then now would probably be the easiest time to do it.

**Review by tomponline — APPROVED** [review]

File: lxd/cluster/gateway.go (hunk 2 above)
Comment (tomponline) [inline lxd/cluster/gateway.go:206]: @monstermunchkin why is the nodeName argumet not populated here?

Comment (monstermunchkin) [reply lxd/cluster/gateway.go:206]: Sounds like a good idea as the warning type itself (i.e. the string representation) says what it's about.

Comment (monstermunchkin) [reply lxd/cluster/gateway.go]: Fixed

**Comment by monstermunchkin** [comment]
jenkins: test this please

**Pull request merged**